recreate docker init layer files

bergwolf opened the pull request:
So that it is possible for the container creation code to skip
the init layer all together.

laijs commented on src/container.c:
```diff
@@ -224,6 +224,55 @@ static int container_setup_mount(struct hyper_container *container)
+
+	if (container_recreate_file("/.dockerenv") < 0)
+		return -1;
+
```
`"/.dockerinit"` and `"/.dockerenv"` were introduced for lxc driver in docker,
`"/.dockerinit"` was already removed, `"/.dockerenv"` might be removed in future,
or might be used for other purpose.

we don't need these files.

laijs commented on src/container.c:
```diff
@@ -224,6 +224,55 @@ static int container_setup_mount(struct hyper_container *container)
+
+	if (container_recreate_symlink("/proc/mounts", "/etc/mtab") < 0)
+		return -1;
+
```
before this patch, we don't have it(/etc/mtab), we need to add it.

laijs commented on src/util.c:
```diff
@@ -133,6 +133,21 @@ int hyper_find_sd(char *prefix, char *addr, char **dev) {
+{
+	return creat(hyper_path, 0755);
+}
+
```
they are so simple, we can move them in util.h

laijs commented on src/container.c:
```diff
@@ -487,6 +536,11 @@ static int hyper_container_init(void *data)
+		fprintf(stderr, "container sets up init layer failed\n");
+		goto fail;
+	}
+
```
it should be called before container_setup_mount() which uses the created mountpoints.

bergwolf commented:
@laijs thanks for reviewing! I've updated the patch to address your comments, and also rebased to latest master to adapt your earlier change on oldrootfs.

laijs commented on src/container.c:
```diff
@@ -224,6 +224,55 @@ static int container_setup_mount(struct hyper_container *container)
+	hyper_unlink(filename);
+	return hyper_create(filename);
+}
+
```
hyperstart already create mountpoint when it need to mount something.

but container_recreate_file() might be still need.

current hyperstart create the mountpoints without unlink them at first.
it will have problems when the user intently create the mountpoint in they image at first.
1) file type dismatch (fail to mount)
2) file permission dismatch (user process might fail)

but these hardly happen (when the user intently create the wrong
mountpoint in they image at first.) But IMO, it is the user's responsibility for the fail.
and in most cases, the container doesn't touch the rootfs(`hyper run ubuntu ls`),
and the `unlink` add burden to the storage driver (especially rbd driver).

I think we can check the files, if the file type & mode match we don't unlink them?

bergwolf replied:
The init layer is used by docker to over ride those files/dirs in the image, from container layer point of view. If performance is a concern here, yes we can skip unlink in some cases, such as those files exist and are empty, providing type and mode match.

Actually, docker would simply unlink, then stat and only create if stat fails with ENOENT. So to have similar behaviour and minimal performance impact, we can do stat first and skip unlink/create if file size is 0, otherwise we do unlink + create.

laijs commented:
I copy one of my note in the code here:

hyperstart already create mountpoint when it need to mount something.

but container_recreate_file() might be still need with some changes.

current hyperstart creates the mountpoints without unlink them at first.
it will have problems when the user intently create the mountpoint in they image at first.
1) file type dismatch (fail to mount)
2) file permission dismatch (user process might fail)

but these hardly happen (only when the user intently create the wrong
mountpoint in they image at first.) But IMO, it is the user's responsibility for the fail.
and in most cases, the container doesn't touch the rootfs(`hyper run ubuntu ls`),
and the `unlink` adds burden to the storage driver (especially rbd driver).

I think we can check the files, if the file type & mode match we don't unlink them?

laijs commented on src/container.c:
```diff
@@ -200,6 +200,49 @@ static int container_setup_mount(struct hyper_container *container)
+
+	if (container_recreate_symlink("./proc/mounts", "./etc/mtab") < 0)
+		return -1;
+
```
it should be "/proc/mounts", it is the content of the symlink.

bergwolf replied:
good catch!

bergwolf commented:
I updated the patch to reflect above discussion.

laijs commented on src/container.c:
```diff
@@ -200,6 +200,61 @@ static int container_setup_mount(struct hyper_container *container)
+ * Docker also creates directories like /dev/pts, /dev/shm, /proc, /sys,
+ * which we have over ridden in container_setup_mount, so no need to create
+ * them here.
+ */
```
docker also unlink /proc, /sys, unlink the any prefix(parent dir) on the path if the prefix is non-dir.

https://github.com/shykes/docker/pull/36

IMO, the unlink is not really necessary, we will polish it later.

laijs commented:
LGTM
